modules/din_upstreams.go

```
@@ -20,11 +24,92 @@ var (
 type DinUpstreams struct{}
 
 type upstreamWrapper struct {
-	HttpUrl  string `json:"http.url"`
-	path     string
-	Headers  map[string]string
-	upstream *reverseproxy.Upstream
-	Priority int
+	HttpUrl   string `json:"http.url"`
+	path      string
+	Headers   map[string]string
+	upstream  *reverseproxy.Upstream
+	Priority  int
+	HCRPCMethod string `json:"healthcheck.rpc.method"`
+	failures  int
+	successes int
+	healthy   bool
+	quit      chan struct{}
+}
+
+func (u *upstreamWrapper) Available() bool {
+	return u.upstream.Available() && u.Healthy()
+}
+
+func (u *upstreamWrapper) StartHealthchecks() {
+	u.quit = make(chan struct{})
+	c := &http.Client{Transport: &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		MaxIdleConnsPerHost:   16,
+		MaxIdleConns:          16,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}}
+	u.healthy = true
+	ticker := time.NewTicker(time.Second * 5)
+	go func() {
+		for i :=0 ; ; i++ {
+			select {
+			case <-u.quit:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				rpcCallString := fmt.Sprintf(`{"jsonrpc":"2.0","id":%v,"method":"%v","params":[]}`, i, u.HCRPCMethod)
+				req, err := http.NewRequest("POST", u.HttpUrl, strings.NewReader(rpcCallString))
+				if err != nil {
+					u.markFailure(err.Error())
```

i feel like this wouldn't be a criteria to mark the proxy for error but instead to send an error via a channel or something. I'm not sure. It probably isn't that big of a deal though.

---

Looking at https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/net/http/request.go;l=855

The only way NewRequest returns an error is if: 

* The method is invalid (but "POST" is hard coded and always valid)
* The URL is invalid (but u.HttpUrl was validated in the middleware's unmarshaller)

Since neither of these conditions can be met given the current codebase, I could make a case for ignoring the error entirely. My concern is that as we start to integrate with the smart contract we may not do as much validation of the URL, and at the very least having the upstream get marked as unhealthy will likely save us some grief.

Once we integrate with the authentication stuff (which passes logging around) we'll have logs of the errors, so if something is unhealthy because of an invalid URL, that should become reasonably obvious from the logs that markFailure will emit.

---

good point. i think in this case you're right that marking it as unhealthy is probably better than just throwing an error.